### PR TITLE
Show full derivation chain when encountering build failures

### PR DESCRIPTION
**Pull request by charliermarsh**
## Summary

This PR adds context to our error messages to explain _why_ a given package was included, if we fail to download or build it.

It's quite a large change, but it motivated some good refactors and improvements along the way.

Closes https://github.com/astral-sh/uv/issues/8962.

### CODE REVIEW

**Review by zanieb — APPROVED** [review]
This looks great; a review from @BurntSushi seems prudent since I'm not as familiar with the resolver.

File: crates/uv-requirements/src/lib.rs [hunk 1]
```diff
@@ -38,6 +49,22 @@ pub enum Error {
...
+            Dist::Source(dist) => {
+                if dist.is_local() {
+                    Self::Build(Box::new(dist), DerivationChain::default(), cause)
+                } else {
+                    Self::DownloadAndBuild(Box::new(dist), DerivationChain::default(), cause)
+                }
+            }
+        }
+    }
+}
```
Comment (BurntSushi) [inline crates/uv-requirements/src/lib.rs:66]: These new helper constructors are a nice refactor!

File: crates/uv-resolver/src/resolver/derivation.rs [hunk 2]
```diff
@@ -0,0 +1,137 @@
...
+                return false;
+            };
+            target == dist.as_ref()
+        })?;
+
+        // Perform a BFS to find the shortest path to the root.
+        let mut queue = VecDeque::new();
+        queue.push_back((target, Vec::new()));
+
+        // TODO(charlie): Consider respecting markers here.
```
Comment (BurntSushi) [inline crates/uv-resolver/src/resolver/derivation.rs:45]: Without markers, I think this means that you might find the shortest path, but that path might not be possible? Or is there another downside I'm not seeing?

File: crates/uv-resolver/src/resolver/derivation.rs [hunk 3]
```diff
@@ -0,0 +1,137 @@
...
+                            } else {
+                                // If we've reached the root, return.
+                                return true;
+                            }
+                        }
+                    }
+                }
+            }
+            false
+        }
```
Comment (BurntSushi) [inline crates/uv-resolver/src/resolver/derivation.rs:122]: This is neat.

File: crates/uv/tests/it/lock.rs [hunk 4]
```diff
@@ -19223,3 +19227,178 @@ fn lock_dynamic_version() -> Result<()> {
...
+            File "[CACHE_DIR]/[TMP]/src/ez_setup/__init__.py", line 170
+              print "Setuptools version",version,"or greater has been installed."
+              ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+          SyntaxError: Missing parentheses in call to 'print'. Did you mean print(...)?
+
+      help: `wsgiref` was included because `project==0.1.0` depends on `wsgiref`
+    "###);
+
+    Ok(())
+}
```
Comment (BurntSushi) [inline crates/uv/tests/it/lock.rs:19404]: Maybe I missed it, but do you have a test for the case where the failed dependency is not a direct dependency of the root? I was looking for that to see what the error message looks like.

**Review by BurntSushi — APPROVED** [review]
I like the refactor and this seems like an overall great UX improvement too!

File: crates/uv/tests/it/lock.rs (hunk 4 above)
Comment (zanieb) [reply crates/uv/tests/it/lock.rs:20007]: I also was hoping for a longer chain :)

Comment (charliermarsh) [reply crates/uv/tests/it/lock.rs:20007]: It's so annoying to test, but yeah, I definitely need to do it.

**Pull request merged**